### ISSUE TITLE
Only MoE experts Fuse test

### ISSUE THREAD

**Issue by yeliang2258** [UNITES-Lab/MC-SMoE#2]
Hello, Thank you for sharing your work!  
I would like to ask, if I only perform expert merging without compression and further training, how significant is the decline in model performance? Are there any relevant tests available? Thanks!

**Comment by pingzhili** [comment]
Thanks for your interest in our work! 

There will be some performance degradation, depending on the specific task. For instance, our prior experiments, _with just expert merging and no further training for recovery_, show that the **SST2** model may experience a slight accuracy decrease of about 1-2%, whereas the **SQuAD** model could see a reduction of approximately 10-20% in the F1 score.

To verify these results yourself, simply execute the merge command as outlined in the README, ensuring to set the `no_eval_until_epochs` argument to 0, and check the step 0 evaluation performance.

**Comment by yeliang2258** [comment]
Thank you for your reply. I appreciate your assistance. 
I have another question: I saw the expert merge code for the Mixtral model, but I didn't see any permute. Is it not necessary to permute in the Mixtral model?

**Comment by pingzhili** [comment]
Intuitively, for sure permutation alignment would help Mixtral expert merging. 

We haven't released this part of the code due to **two reasons**: 

**1)** Mixtral employs a different FFN architecture - SwiGLU - and the original permutation alignment algorithm may not be optimal for this configuration. We were trying to develop optimal ones but did not much progress due to computing limits. 

**2)** While permutation alignment is designed for addressing merging issues in neural networks that are from different initialization (like typical MoE experts), rumors said Mixtral was trained from dense upcycling and then its MoE experts are from the same initialization if that is true.

We will release a script for Mixtral with the original permutation alignment algorithm soon, should you be interested.

**Comment by yeliang2258** [comment]
Thank you for your detailed response and assistance. I am very interested in the alignment algorithm you mentioned. Could you please share it with me when it becomes available? I appreciate your help!

**Comment by yeliang2258** [comment]
While reviewing the `grouping_mixtral.py` code, I have a question regarding the `group_experts_globally_from_dominant_experts` function: Should the line `most_similar_core = core_expert_indices[torch.argmax(i, core_expert_indices)]` be changed to `most_similar_core = core_expert_indices[torch.argmax(similarity_matrix[indices_sorted_by_usage[i], core_expert_indices])]`?
```
 # 2 Assign left unassigned experts to the cluster with the most similar core
            similarity_matrix = self.get_similarity_matrix(ffn_name)
            for i in range(num_groups, self.num_experts):
                # Find the most similar core
                most_similar_core = core_expert_indices[
                    torch.argmax(similarity_matrix[i, core_expert_indices]) 
                ]
                most_similar_group_label = self._group_state_dict[ffn_name][most_similar_core]
                self._group_state_dict[ffn_name][i] = most_similar_group_label
```
Fix to ?
```
 # 2 Assign left unassigned experts to the cluster with the most similar core
            similarity_matrix = self.get_similarity_matrix(ffn_name)
            for i in range(num_groups, self.num_experts):
                # Find the most similar core
                most_similar_core = core_expert_indices[
                    torch.argmax(similarity_matrix[indices_sorted_by_usage[i], core_expert_indices]) # bug?????
                ]
                most_similar_group_label = self._group_state_dict[ffn_name][most_similar_core]
                self._group_state_dict[ffn_name][indices_sorted_by_usage[i]] = most_similar_group_label # bug?????
```

**Comment by yeliang2258** [comment]
非常感谢分享，我这边还有一个问题，Mixtral 的 merge，其实没有修改router，只是将experts的权重合并了，同一个组的专家共享权重，这样的模型对部署来说并不容易，请问一下，如果我同一个组的专家就保留一个，相应的也减小router，那么router应该怎么做剪枝或者合并呢？你们有相关的经验或者实验吗？

**Comment by pingzhili** [comment]
Yes, this should be a bug presented a long time ago... It was fixed before we got the results in our paper, but somehow we have uploaded the elder version of this script. Anyway, thanks for bringing this up and we will upload the correct one soon.

**Comment by pingzhili** [comment]
The router merging is a good open question. Our initial experiments on SwitchTransformers, by simply averaging the router rows, had shown suboptimal performance. 

Another idea that may be worth a try is to employ some advanced new routers (e.g. MLP), and train them to learn the original routing patterns but with reduced expert numbers.

**Comment by yeliang2258** [comment]
I used the [msmoe-merging-mixtral.py](http://msmoe-merging-mixtral.py/) script to test the expert Fuse method, and the results are as follows. It seems that the PPL (Perplexity) increases rapidly. Is this expected?

Expert Num |  wikitext2 PPL | C4 PPL
-- | -- | --
8 | 3.9268 | 6.87911
6 | 29.4664 | 38.4368
4 | 5013.6641 | 3600.1140

**Comment by pingzhili** [comment]
I think it is reasonable without post-merging fine-tuning, I also expect the performance could be recovered soon with not too many steps of training.

**Comment by yeliang2258** [comment]
Thank you for your response. Thank you!
